src/components/database/updateReviewerStats.ts

```diff
@@ -0,0 +1,231 @@
+  reviewerIds = reviewerIds.filter((id) => id !== null)
+
+  for (const userId of reviewerIds) {
+    console.log(`Updating actions for user ${userId} on application ${applicationId}`)
+    await db.upsertReviewerAction(applicationId, userId)
+  }
+
+  // Clean up NULL records
+  await db.cleanupNullReviewerActions(applicationId)
+}
+
+// When triggered by a Database event, this preliminary function gets the
+// applicationId and the list of reviewers to update. It is different for each
+// trigger case, so this function can standardise the input required by
+// `updateReviewerStats`, which does the actual updates
+export const updateReviewerStatsFromDBEvent = async ({
+  tableName,
+  data,
+  operation,
+}: NotificationPayload) => {
+  let applicationId: number = 0
+  let reviewerIds: number[] = []
+
+  const db = databaseMethods
+
+  console.log(
+    `\nUpdating reviewer action stats due to ${operation} trigger on table: ${tableName} `
+  )
+
+  switch (tableName) {
+    case 'review_assignment':
+      applicationId = data.application_id as number
+      // Update ALL staff by leaving reviewerIds empty
+      break
+    case 'review_status_history':
+      const { application_id, reviewer_id, assigner_id } = await db.getApplicationDataFromReview(
+        data.review_id as number
+      )
+      applicationId = application_id
+      reviewerIds.push(...new Set([reviewer_id, assigner_id]))
+      break
+    case 'application':
+      applicationId = data.id as number
+      // Update ALL staff by leaving reviewerIds empty
+      break
+  }
+  updateReviewerStats(applicationId, reviewerIds.length === 0 ? undefined : reviewerIds)
+}
+
+const databaseMethods = {
+  updateApplicationStaffLists: async (appId: number) => {
+    try {
+      const text = `
+        WITH lists AS (
+                SELECT reviewers, assigners
+                FROM single_application_detail($1)
+            ) 
+        UPDATE public.application
+            SET reviewer_list = (SELECT reviewers FROM lists),
+            assigner_list = (SELECT assigners FROM lists)
+            WHERE id = $1;`
+      await DBConnect.query({
+        text,
+        values: [appId],
+      })
+    } catch (err) {
+      console.log('ERROR updating staff lists for application', appId)
+      console.log(errorMessage(err))
+      throw err
+    }
+  },
+  getAllStaffForApplication: async (appId: number) => {
+    try {
+      const text = `
+        SELECT DISTINCT reviewer_id FROM review_assignment
+            WHERE application_id = $1
+        UNION 
+        SELECT DISTINCT assigner_id from public.review_assignment_assigner_join
+            WHERE review_assignment_id IN (
+                SELECT id FROM public.review_assignment
+                WHERE application_id = $1
+            );
+        `
+      const result = await DBConnect.query({
+        text,
+        values: [appId],
+        rowMode: 'array',
+      })
+      return result.rows.flat()
```

Comment:
it's possible that reviewers will also be assigners ? I think we would want to dedup

Reply:
The `DISTINCT` in each subquery and the `UNION` results in a list of unique IDs, no dupes (assigners and reviewers combined). You would think the `UNION` would combine them, but it doesn't in this case 🤷‍♂️

Reply:
Ohh true, looks like union does distict on rows https://www.postgresql.org/docs/current/queries-union.html